core/solver/cgs.cpp

```diff
@@ -0,0 +1,218 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright 2017-2018
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/solver/cgs.hpp"
+
+
+#include "core/base/convertible.hpp"
+#include "core/base/exception.hpp"
+#include "core/base/exception_helpers.hpp"
+#include "core/base/executor.hpp"
+#include "core/base/math.hpp"
+#include "core/base/utils.hpp"
+#include "core/solver/cgs_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+
+
+namespace {
+
+
+template <typename ValueType>
+struct TemplatedOperation {
+    GKO_REGISTER_OPERATION(initialize, cgs::initialize<ValueType>);
+    GKO_REGISTER_OPERATION(step_1, cgs::step_1<ValueType>);
+    GKO_REGISTER_OPERATION(step_2, cgs::step_2<ValueType>);
+    GKO_REGISTER_OPERATION(step_3, cgs::step_3<ValueType>);
+    GKO_REGISTER_OPERATION(step_4, cgs::step_4<ValueType>);
+};
+
+
+/**
+ * Checks whether the required residual goal has been reached or not.
+ *
+ * @param tau  Residual of the iteration.
+ * @param orig_tau  Original residual.
+ * @param r  Relative residual goal.
+ */
+template <typename ValueType>
+bool has_converged(const matrix::Dense<ValueType> *tau,
+                   const matrix::Dense<ValueType> *orig_tau,
+                   remove_complex<ValueType> r)
```

Comment:
I think this is exactly the same as `CG` for example. Is there a way to share this function, or maybe there is a reason for not doing that?

Reply:
@gflegar ?

Reply:
This is a known issue, see https://gitlab.com/ginkgo-project/ginkgo/issues/60

Reply:
Maybe to expand on this: yes we could remove this duplication, but a true fix is to add another class hierarchy for different kinds of convergence tests.  We just haven't gotten around to doing that yet.

Reply:
True I had forgotten this issue (and related 55). Anyway since we'll change that at some point, let's not lose time on this for now.